File: C++/VendingMachine/VendingMachine.UnitTestsG/CreditShould.cpp
#include "pch.h"
#include "../VendingMachine/Credit.h"
#include "../VendingMachine/InsufficientCreditError.h"

TEST(CreditShould, ReturnZeroValueWithNoCreditAdded)
{
	Credit credit;

	EXPECT_EQ(credit.GetValue(), 0);
}

TEST(CreditShould, ReturnValueEqualToSingleCreditAdded)
{
	const double expectedCreditValue = 1;

	const double testCreditValue = 1;

	Credit credit;
	credit.Add(testCreditValue);

	const auto actualCreditValue = credit.GetValue();
	
	EXPECT_EQ(actualCreditValue, expectedCreditValue);
}

TEST(CreditShould, ReturnValueEqualToSumOfMultipleCreditsAdded)
{
	const double expectedCreditValue = 2;

	const double testCreditValue = 1;

	Credit credit;
	credit.Add(testCreditValue);
	credit.Add(testCreditValue);

	const auto actualCreditValue = credit.GetValue();
	
	EXPECT_EQ(actualCreditValue, expectedCreditValue);
}

TEST(CreditShould, ThrowExceptionIfSpendIsGreaterThanTotalValue)
{
	const double testSpendAmount = 1;

	Credit credit;

	EXPECT_THROW(credit.Spend(testSpendAmount), InsufficientCreditError);
}

TEST(CreditShould, ReturnFalseWhenValidatingSpendWithInsufficientCredit)
{
	const double testSpendAmount = 1;

	const Credit credit;
	const auto actualResult = credit.CanSpend(testSpendAmount);

	EXPECT_FALSE(actualResult);
}

TEST(CreditShould, ReturnZeroValueAfterSpendEqualToTotalValue)
{
	const double expectedRemainingValue = 0;

	const double testCreditValue = 1;
	const double testSpendValue = 1;

	Credit credit;
	credit.Add(testCreditValue);
	credit.Spend(testSpendValue);

	auto actualRemainingValue = credit.GetValue();

	EXPECT_EQ(actualRemainingValue, expectedRemainingValue);
}

TEST(CreditShould, ReturnRemainingValueAfterSpendLessThanTotalValue)
{
	const double expectedRemainingValue = 1;

	const double testCreditValue = 2;
	const double testSpendValue = 1;

	Credit credit;
	credit.Add(testCreditValue);
	credit.Spend(testSpendValue);

	auto actualRemainingValue = credit.GetValue();

	EXPECT_EQ(actualRemainingValue, expectedRemainingValue);
}

File: java/VendingMachine/test/com/ravendarque/unitTests/CreditShould.java
package com.ravendarque.unitTests;

import com.ravendarque.vendingMachine.credit.Credit;
import com.ravendarque.vendingMachine.credit.InsufficientCreditException;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;

class CreditShould {

    @Test
    void returnZeroValueWithNoCreditAdded() {

        final double expectedCreditValue = 0;

        final Credit credit = new Credit();

        final double actualCreditValue = credit.getValue();

        assertEquals(expectedCreditValue, actualCreditValue);
    }

    @Test
    void returnValueEqualToSingleCreditAdded() {

        final double expectedCreditValue = 2;
        final double creditToAdd = 2;

        final Credit credit = new Credit();
        credit.add(creditToAdd);

        final double actualCreditValue = credit.getValue();

        assertEquals(expectedCreditValue, actualCreditValue);
    }

    @Test
    void returnValueEqualToSumOfMultipleCreditsAdded() {

        final double expectedCreditValue = 4;
        final double creditToAdd = 2;

        final Credit credit = new Credit();
        credit.add(creditToAdd);
        credit.add(creditToAdd);

        final double actualCreditValue = credit.getValue();

        assertEquals(expectedCreditValue, actualCreditValue);
    }

    @Test
    void throwExceptionIfSpendIsGreaterThanTotalValue() {

        final double testSpendAmount = 2;
        final double testCreditValue = 1;

        final Credit credit = new Credit();
        credit.add(testCreditValue);

        assertThrows(InsufficientCreditException.class, () -> credit.spend(testSpendAmount));
    }

    @Test
    void returnFalseWhenValidatingSpendWithInsufficientCredit() {

        final boolean expectedResult = false;
        final int testSpendAmount = 2;

        final Credit credit = new Credit();

        final boolean actualResult = credit.canSpend(testSpendAmount);

        assertEquals(expectedResult, actualResult);
    }

    @Test
    void returnZeroValueAfterSpendEqualToTotalValue()
            throws InsufficientCreditException {

        final double expectedRemainingValue = 0;

        final int testCreditValue = 1;
        final int testSpend = 1;
        final Credit credit = new Credit();
        credit.add(testCreditValue);
        credit.spend(testSpend);

        final double actualRemainingValue = credit.getValue();

        assertEquals(expectedRemainingValue, actualRemainingValue);
    }

    @Test
    void returnRemainingCreditAfterSpendLessThanTotalValue()
            throws InsufficientCreditException {

        final double expectedRemainingValue = 1.5;
        final int testCreditValue = 3;
        final double testSpendAmount = 1.5;

        final Credit credit = new Credit();
        credit.add(testCreditValue);
        credit.spend(testSpendAmount);

        final double actualRemainingValue = credit.getValue();

        assertEquals(expectedRemainingValue, actualRemainingValue);
    }
}
